windows libusb: show only winusb devices

karel-3d opened the pull request:
Libusb lists all devices, even those without drivers.

What can happen is that a user can open web wallet, connect a new device, and while the winusb driver is installing, the device is shown to wallet, which tries to immediately connect to it and errors. User sees libusb error and thinks something is broken (while reconnecting device would have actually helped).

So what we need to do is to exclude devices without winusb drivers in device listing.

I was too afraid to change the device listing in os/windows_winusb.c to exclude it directly there - because I was afraid of a memory leak (the memory is manually managed for the device objects).

Instead, I did a dirtier solution - I have added a field to the device struct that is by default 0 on both composite devices and "single" devices, but is 1 if either the "single" device is winusb or if any of the sub-devices on the composite device is winusb. This field is then looked at in device listing *in golang layers*.

This could MAYBE cause an issue some day in the future if we have MORE winusb endpoints, and this will get set as 1 on the first endpoint, even when the other endpoints won't be inited. However this would cause more havoc probably anyway in the golang layers, and also probably in webusb in chrome. So, it is not really an issue I guess

I can confirm this fixes the issue on my windows 7 - the device is not displayed in wallet while it the device is installed, it appears in wallet immediately after the installation is done.

(PS: I have no idea how did Chromium solve this with their heavily patched libusb fork... because they probably have the same issue.)

karel-3d commented:
It seems I have broken Linux build :( I will look at it

prusnak commented on usb/webusb.go:
```diff
@@ -197,6 +197,11 @@ func (b *WebUSB) connect(dev lowlevel.Device) (*WUD, error) {
 }
 
 func (b *WebUSB) match(dev lowlevel.Device) bool {
+	if isWindows {
```
Maybe it's easier to just check whether `runtime.GOOS` is `windows/386` or `windows/amd64` rather than introducing 2 new files. But this is just a nitpick.

karel-3d replied:
I have read somewhere that it is not so dependable and compile time is better

karel-3d replied:
... but you are probably right. I will edit the commit

karel-3d replied:
OK, I will edit the commit, you are probably right

karel-3d commented:
Fixed the "go get" issues.... from the second commit:

Because I added libusbi.h to libusb.go, go stopped building it because of some cgo magic.

I have to admit I don't understand cgo that much. But when I shuffled around the headers it started to build again.

prusnak commented on usb/webusb.go:
```diff
@@ -196,7 +196,19 @@ func (b *WebUSB) connect(dev lowlevel.Device) (*WUD, error) {
 }
 
+func isWindows() bool {
+	if runtime.GOOS == "windows/386" || runtime.GOOS == "windows/amd64" {
```
Maybe just `return if runtime.GOOS == "windows/386" || runtime.GOOS == "windows/amd64"` ?
